st2common/st2common/util/green/shell.py

```diff
@@ -80,29 +80,30 @@ def run_command(cmd, stdin=None, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
     process = subprocess.Popen(args=cmd, stdin=stdin, stdout=stdout, stderr=stderr,
                                env=env, cwd=cwd, shell=shell, preexec_fn=preexec_func)
 
-    def on_timeout_expired(timeout):
-        global timed_out
-
-        try:
-            process.wait(timeout=timeout)
-        except subprocess.TimeoutExpired:
-            # Command has timed out, kill the process and propagate the error.
-            # Note: We explicitly set the returncode to indicate the timeout.
-            process.returncode = TIMEOUT_EXIT_CODE
-
-            if kill_func:
-                kill_func(process=process)
-            else:
-                process.kill()
-
-    timeout_thread = eventlet.spawn(on_timeout_expired, timeout)
-    stdout, stderr = process.communicate()
-    timeout_thread.cancel()
-    exit_code = process.returncode
-
-    if exit_code == TIMEOUT_EXIT_CODE:
-        timed_out = True
+    if not timeout:
+        stdout, stderr = process.communicate()
```

Comment:
We achieve true parallelism by asking users to run multiple workers at the same time. Each worker takes a message from the queue and does use a green thread to process the message (kicking off the execution). So yeah, you are right that it should just block a thread with infinite timeout. But if you run multiple such actions, you'd pretty much start blocking the pool.

I also think there's a bug in our current implementation that we ack early. See https://github.com/StackStorm/st2/blob/master/st2common/st2common/transport/consumers.py#L52. I am afraid this might mean we tell st2 that action is done or kicked off while it might still be in the dispatcher queue waiting for an eventlet to work on. We are probably not hitting the case because pool size is 50. This is a different issue but your question is tangentially related.